File: typescript/packages/e2b/src/driver.ts
import type { Fetch, SandboxDriver } from "@threads/core/adapter";
import { type Envd, envd, envdUrl } from "./envd";
import { control } from "./rest";
import { sender } from "./transport";
import { type Described, E2bError } from "./wire";

// The remote kit's driver over E2B's REST API and each sandbox's envd, spoken directly
// (rest.ts, envd.ts) so every request leaves through the fenced fetch. The API key only
// authenticates the host's own control-plane requests; a sandbox is created with no env vars,
// and nothing from the host env reaches it (AGENTS invariant 4).

export { OPERATION_KEY } from "./rest";

export type DriverOptions = {
  /** Read per call: the key is resolved at setup, after the factory. */
  readonly apiKey: () => string;
  readonly domain: string;
  readonly template: string;
  readonly timeoutMs: number;
  readonly internet: boolean;
  readonly fetch: Fetch;
};

export function e2bDriver(options: DriverOptions): SandboxDriver {
  const send = sender(options.fetch);
  const rest = control(send, `https://api.${options.domain}`, options.apiKey);
  const timeoutS = Math.max(1, Math.floor(options.timeoutMs / 1000));
  const envds = new Map<string, Envd>();
  const remember = (sandbox: Described): Envd => {
    const url = envdUrl(sandbox.sandboxID, sandbox.domain ?? options.domain);
    const made = envd(send, url, sandbox);
    envds.set(sandbox.sandboxID, made);
    return made;
  };
  /** A live sandbox's envd; after a restart, found again by describing the sandbox. */
  const envdOf = async (id: string): Promise<Envd> => {
    const known = envds.get(id);
    if (known !== undefined) return known;
    const described = await rest.describe(id);
    if (described === null)
      throw new E2bError("not_found", `E2B has no sandbox ${id}`);
    return remember(described);
  };

  return {
    create: async (key, snapshot) => {
      const template = snapshot ?? options.template;
      const made = await rest.create(template, key, {
        timeoutS,
        internet: options.internet,
      });
      if (made !== null) {
        remember(made);
        return { kind: "created", id: made.sandboxID };
      }
      const message = `E2B has no template ${template}`;
      if (snapshot === undefined) throw new E2bError("unavailable", message);
      return { kind: "snapshot_missing", message };
    },
    find: async (key) => {
      const ids = await rest.find(key);
      const [only, ...more] = ids;
      if (only === undefined) return { status: "not_found_nonfinal" };
      if (more.length > 0)
        return {
          status: "unknown",
          reason: `several sandboxes carry ${key}: ${ids.join(", ")}`,
        };
      return { status: "found", value: only };
    },
    exists: async (id) => {
      const described = await rest.describe(id);
      if (described !== null) remember(described);
      return described !== null;
    },
    kill: async (id) => {
      envds.delete(id);
      return (await rest.kill(id)) ? "killed" : "already_gone";
    },
    run: async (id, script, sinks, processKey) =>
      (await envdOf(id)).start(
        { argv: ["/bin/sh", "-c", script], env: {}, cwd: "/", tag: processKey },
        sinks,
      ),
    stopProcess: async (id, processKey) => {
      await (await envdOf(id)).signal(processKey);
    },
    write: async (id, path, data) => (await envdOf(id)).upload(path, data),
    read: async (id, path) => (await envdOf(id)).download(path),
    snapshot: {
      // E2B's snapshot pauses the sandbox, but E2B doesn't document that the pause is a
      // barrier for every descendant and clone, so quiescence stays unconfirmed and the kit
      // takes no snapshots until a live qualification proves it. A restore from an E2B
      // snapshot id still works, verified against the manifest hash.
      quiescence: "unconfirmed",
      // E2B keeps a snapshot until it is deleted.
      take: async (id, key) => ({
        ref: await rest.snapshot(id, `threads-${key}`),
        expiresAt: null,
      }),
    },
    deleteSnapshot: async (ref) =>
      (await rest.deleteTemplate(ref)) ? "released" : "already_gone",
  };
}

File: python/src/threads/adapters/sandboxes/e2b/driver.py
"""The remote kit's driver over E2B (sandbox/remote/driver.py): the control plane for create,
lookup and kill, and each sandbox's envd for exec, files and the process kill, every call on
one event loop's fenced transports (plane.py)."""

from collections.abc import Callable, Mapping, Sequence
from dataclasses import dataclass
from typing import Literal, Self

import httpx
from connectrpc.code import Code
from connectrpc.errors import ConnectError
from pydantic import ValidationError

from threads.adapters.sandboxes.e2b.control import ApiError, MalformedError
from threads.adapters.sandboxes.e2b.envd import Envd
from threads.adapters.sandboxes.e2b.plane import Plane
from threads.adapters.sandboxes.e2b.wire import Sandbox as Described
from threads.adapters.sandboxes.fence import Classify
from threads.adapters.sandboxes.streams import StreamLostError
from threads.loop.model import Found, LookupUnknown, NotFoundNonfinal
from threads.sandbox.protocol import ExecOutput, SandboxError
from threads.sandbox.remote.driver import Capture, NonfinalLookup, Unconfirmed, Unmade

_UNAVAILABLE = (
    ApiError,
    MalformedError,
    ValidationError,
    httpx.TransportError,
    StreamLostError,
    ConnectError,
)
_NOT_FOUND = 404


def classify(error: Exception) -> SandboxError | None:
    """The expected failures of an E2B call; anything else is a bug and raises."""
    deadline = isinstance(error, ConnectError) and error.code == Code.DEADLINE_EXCEEDED
    if deadline or isinstance(error, httpx.TimeoutException):
        return SandboxError("timeout", str(error))
    if isinstance(error, _UNAVAILABLE):
        return SandboxError("unavailable", str(error))
    return None


@dataclass(frozen=True, slots=True)
class Settings:
    template: str
    timeout_s: int
    """E2B's sandbox timeout: it kills the sandbox this long after create."""
    internet: bool


class E2BDriver:
    def __init__(self, plane: Callable[[], Plane], settings: Settings) -> None:
        self._plane, self._settings = plane, settings
        # A create that passed its fence may still be in flight when the query runs.
        self.lookup: NonfinalLookup = NonfinalLookup(self._find)
        # envd kills the process it started, not what that process detached.
        self.termination: Unconfirmed = Unconfirmed(self._signal)
        self.capture: Capture | None = None

    def bound(self) -> Self:
        plane = self._plane()
        return type(self)(lambda: plane, self._settings)

    @property
    def classify(self) -> Classify:
        return classify

    async def create(self, operation_key: str, snapshot: str | None) -> str | Unmade:
        """From the template: with no capture, the kit never restores (`snapshot` is None)."""
        s = self._settings
        control = self._plane().control
        made = await control.create(
            s.template, operation_key, timeout_s=s.timeout_s, internet=s.internet
        )
        if made is None:
            return Unmade("snapshot_missing", f"E2B has no template {s.template}")
        self._open(made)
        return made.sandbox_id

    async def exists(self, sandbox_id: str) -> bool:
        described = await self._plane().control.describe(sandbox_id)
        if described is not None:
            self._open(described)
        return described is not None

    async def kill(self, sandbox_id: str) -> Literal["killed", "already_gone"]:
        """Kills the sandbox, then closes its envd clients."""
        plane = self._plane()
        killed = await plane.control.kill(sandbox_id)
        envd = plane.envds.get(sandbox_id)
        if envd is not None:
            await envd.aclose()
        return "killed" if killed else "already_gone"

    async def run(
        self,
        sandbox_id: str,
        argv: Sequence[str],
        env: Mapping[str, str],
        cwd: str,
        process_key: str | None,
    ) -> ExecOutput:
        return await (await self._envd(sandbox_id)).start(argv, env, cwd, process_key)

    async def write(self, sandbox_id: str, path: str, data: bytes) -> None:
        await (await self._envd(sandbox_id)).upload(path, data)

    async def read(self, sandbox_id: str, path: str) -> bytes:
        return await (await self._envd(sandbox_id)).download(path)

    async def _find(self, operation_key: str) -> Found[str] | NotFoundNonfinal | LookupUnknown:
        match await self._plane().control.find(operation_key):
            case []:
                return NotFoundNonfinal()
            case [ref]:
                return Found(ref)
            case refs:
                return LookupUnknown(f"several sandboxes carry {operation_key}: {refs}")

    async def _signal(self, sandbox_id: str, process_key: str) -> None:
        await (await self._envd(sandbox_id)).signal(process_key)

    async def _envd(self, sandbox_id: str) -> Envd:
        known = self._plane().envds.get(sandbox_id)
        if known is not None:
            return known
        described = await self._plane().control.describe(sandbox_id)
        if described is None:
            raise ApiError(_NOT_FOUND, f"E2B has no sandbox {sandbox_id}".encode())
        return self._open(described)

    def _open(self, described: Described) -> Envd:
        """The sandbox's envd clients on this plane, opened once (they list themselves in
        `plane.envds` until closed)."""
        plane = self._plane()
        known = plane.envds.get(described.sandbox_id)
        if known is not None:
            return known
        domain = described.domain or plane.config.domain
        url = plane.config.get_sandbox_url(described.sandbox_id, domain)
        return Envd(described, url, plane.transports, plane.envds)
